scripts/verify_toolchain_directive.sh

```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+
+# This script verifies that the value of the toolchain directive in the
+# go.mod files always match that of the .go-version file to ensure that
+# we accidentally don't test and release with differing versions of Go.
+
+set -euo pipefail
+
+source ./scripts/test_lib.sh
+
+target_go_version="${target_go_version:-"$(cat "${ETCD_ROOT_DIR}/.go-version")"}"
+log_info "expected go toolchain directive: go${target_go_version}"
+log_info
+
+out_of_sync="false"
+for mod in $(find . -name 'go.mod'); do
+    toolchain_version="$(go mod edit -json ${mod} | jq -r .Toolchain)"
+    if [[ "go${target_go_version}" != "${toolchain_version}" ]]; then
+        log_error "go toolchain directive out of sync for ${mod}, got: ${toolchain_version}"
+        out_of_sync="true"
+    fi
+done
```

Comment:
The `go` lines in all etcd's `go.mod` files are also automatically bumped if any etcd's dependency bumped its go version. Should we also verify that the version in `go lines` <= `.go-version`?

Reply:
@ahrtr that's a good point, however I'm not sure what the course of action is once the dependency bump PRs are merged. If a `go line` bump occurred and the PR is merged, do we just update the `.go-version` file then? 

My main concern with allowing dependency bumps to also bump the go line is that it can cause an issue with projects like Kubernetes that import `go.etcd.io...` modules. If the go lines of these projects are behind those of the etcd modules, then they will be either forced to bump or they cannot update the dependency till they catch up (https://github.com/kubernetes/kubernetes/issues/123744)

This behaviour is mitigated in go1.23, but till that is released and we can adopt that, it is risky allowing the go line to be bumped at all, especially from a perspective of Kubernetes consuming etcd. For bumps related to security fixes or bugs that we see in etcd, we should probably discuss what the course of action should be, but we should probably we a little more aware for usual maintenance bumps.

Reply:
>  they cannot update the dependency till they catch up 

I think we should follow this. 

For main branch, it isn't a problem because usually we keep using the latest go version (e.g. 1.22.2 for now).

For stable releases (3.5 and 3.4 for now), usually we keep using the latest patch of previous minor version (e.g. 1.21.9 for now). Usually, we don't proactively bump dependencies for etcd stable releases unless there are any CVEs or bugs that affect etcd. If bumping a dependency for a stable release leads to the go version being automatically bumped, 
- check whether there is an alternative. For example, if both 4.20.6 and 4.21.0 of a dependency resolve a CVE, but 4.20.6 requires go1.21.9 and 4.21.0 requires go 1.22.2, then we bump 4.20.6 for etcd stable releases (3.4/3.5).
- if there is no alternative
  - if we have to bump the dependency, then we bump go version for etcd stable releases (3.4/3.5) as well.
  - if we don't have to bump the dependency (_it's unlikely running into this case_), we can hold off for now until we bump go version.


We need to clearly document the behaviour in https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/dependency_management.md#golang-versions

Reply:
@ahrtr that sounds good, I've added the extra check.

Reply:
Also, ack on documenting this.